// itk-wasm/dist/web-workers/bundles/mesh-io.worker.js
(function () {
  'use strict';

  var register = {exports: {}};

  var _createClass$1 = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

  function _classCallCheck$1(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  var TinyEmitter$1 = function () {
    function TinyEmitter() {
      _classCallCheck$1(this, TinyEmitter);

      Object.defineProperty(this, '__listeners', {
        value: {},
        enumerable: false,
        writable: false
      });
    }

    _createClass$1(TinyEmitter, [{
      key: 'emit',
      value: function emit(eventName) {
        if (!this.__listeners[eventName]) return this;

        for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          args[_key - 1] = arguments[_key];
        }

        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = this.__listeners[eventName][Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var handler = _step.value;

            handler.apply(undefined, args);
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        return this;
      }
    }, {
      key: 'once',
      value: function once(eventName, handler) {
        var _this = this;

        var once = function once() {
          _this.off(eventName, once);
          handler.apply(undefined, arguments);
        };

        return this.on(eventName, once);
      }
    }, {
      key: 'on',
      value: function on(eventName, handler) {
        if (!this.__listeners[eventName]) this.__listeners[eventName] = [];

        this.__listeners[eventName].push(handler);

        return this;
      }
    }, {
      key: 'off',
      value: function off(eventName, handler) {
        if (handler) this.__listeners[eventName] = this.__listeners[eventName].filter(function (h) {
          return h !== handler;
        });else this.__listeners[eventName] = [];

        return this;
      }
    }]);

    return TinyEmitter;
  }();

  var tinyEmitter = TinyEmitter$1;

  var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

  var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

  var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

  function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

  function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  var TinyEmitter = tinyEmitter;

  var MESSAGE_RESULT = 0;
  var MESSAGE_EVENT = 1;

  var RESULT_ERROR = 0;
  var RESULT_SUCCESS = 1;

  var DEFAULT_HANDLER = 'main';

  var isPromise = function isPromise(o) {
    return (typeof o === 'undefined' ? 'undefined' : _typeof(o)) === 'object' && o !== null && typeof o.then === 'function' && typeof o.catch === 'function';
  };

  function RegisterPromise(fn) {
    var handlers = _defineProperty({}, DEFAULT_HANDLER, fn);
    var sendPostMessage = self.postMessage.bind(self);

    var server = new (function (_TinyEmitter) {
      _inherits(WorkerRegister, _TinyEmitter);

      function WorkerRegister() {
        _classCallCheck(this, WorkerRegister);

        return _possibleConstructorReturn(this, (WorkerRegister.__proto__ || Object.getPrototypeOf(WorkerRegister)).apply(this, arguments));
      }

      _createClass(WorkerRegister, [{
        key: 'emit',
        value: function emit(eventName) {
          for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            args[_key - 1] = arguments[_key];
          }

          if (args.length == 1 && args[0] instanceof TransferableResponse) {
            sendPostMessage({ eventName: eventName, args: args }, args[0].transferable);
          } else {
            sendPostMessage({ eventName: eventName, args: args });
          }
          return this;
        }
      }, {
        key: 'emitLocally',
        value: function emitLocally(eventName) {
          var _get2;

          for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
            args[_key2 - 1] = arguments[_key2];
          }

          (_get2 = _get(WorkerRegister.prototype.__proto__ || Object.getPrototypeOf(WorkerRegister.prototype), 'emit', this)).call.apply(_get2, [this, eventName].concat(args));
        }
      }, {
        key: 'operation',
        value: function operation(name, handler) {
          handlers[name] = handler;
          return this;
        }
      }]);

      return WorkerRegister;
    }(TinyEmitter))();

    var run = function run(messageId, payload, handlerName) {

      var onSuccess = function onSuccess(result) {
        if (result && result instanceof TransferableResponse) {
          sendResult(messageId, RESULT_SUCCESS, result.payload, result.transferable);
        } else {
          sendResult(messageId, RESULT_SUCCESS, result);
        }
      };

      var onError = function onError(e) {
        sendResult(messageId, RESULT_ERROR, {
          message: e.message,
          stack: e.stack
        });
      };

      try {
        var result = runFn(messageId, payload, handlerName);
        if (isPromise(result)) {
          result.then(onSuccess).catch(onError);
        } else {
          onSuccess(result);
        }
      } catch (e) {
        onError(e);
      }
    };

    var runFn = function runFn(messageId, payload, handlerName) {
      var handler = handlers[handlerName || DEFAULT_HANDLER];
      if (!handler) throw new Error('Not found handler for this request');

      return handler(payload, sendEvent.bind(null, messageId));
    };

    var sendResult = function sendResult(messageId, success, payload) {
      var transferable = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];

      sendPostMessage([MESSAGE_RESULT, messageId, success, payload], transferable);
    };

    var sendEvent = function sendEvent(messageId, eventName, payload) {
      if (!eventName) throw new Error('eventName is required');

      if (typeof eventName !== 'string') throw new Error('eventName should be string');

      sendPostMessage([MESSAGE_EVENT, messageId, eventName, payload]);
    };

    self.addEventListener('message', function (_ref) {
      var data = _ref.data;

      if (Array.isArray(data)) {
        run.apply(undefined, _toConsumableArray(data));
      } else if (data && data.eventName) {
        server.emitLocally.apply(server, [data.eventName].concat(_toConsumableArray(data.args)));
      }
    });

    return server;
  }

  var TransferableResponse = function TransferableResponse(payload, transferable) {
    _classCallCheck(this, TransferableResponse);

    this.payload = payload;
    this.transferable = transferable;
  };

  register.exports = RegisterPromise;
  register.exports.TransferableResponse = TransferableResponse;

  var registerWebworker = register.exports;

  const mimeToIO = new Map([]);
  var mimeToIO$1 = mimeToIO;

  function getFileExtension(filePath) {
      let extension = filePath.slice((filePath.lastIndexOf('.') - 1 >>> 0) + 2);
      if (extension.toLowerCase() === 'gz') {
          const index = filePath.slice(0, -3).lastIndexOf('.');
          extension = filePath.slice((index - 1 >>> 0) + 2);
      }
      else if (extension.toLowerCase() === 'cbor') {
          const index = filePath.slice(0, -5).lastIndexOf('.');
          extension = filePath.slice((index - 1 >>> 0) + 2);
      }
      else if (extension.toLowerCase() === 'zstd') {
          // .iwi.cbor.zstd
          const index = filePath.slice(0, -10).lastIndexOf('.');
          extension = filePath.slice((index - 1 >>> 0) + 2);
      }
      else if (extension.toLowerCase() === 'zip') {
          const index = filePath.slice(0, -4).lastIndexOf('.');
          extension = filePath.slice((index - 1 >>> 0) + 2);
      }
      return extension;
  }

  const extensionToIO = new Map([
      ['vtk', 'itkVTKPolyDataMeshIOJSBinding'],
      ['VTK', 'itkVTKPolyDataMeshIOJSBinding'],
      ['byu', 'itkBYUMeshIOJSBinding'],
      ['BYU', 'itkBYUMeshIOJSBinding'],
      ['fsa', 'itkFreeSurferAsciiMeshIOJSBinding'],
      ['FSA', 'itkFreeSurferAsciiMeshIOJSBinding'],
      ['fsb', 'itkFreeSurferBinaryMeshIOJSBinding'],
      ['FSB', 'itkFreeSurferBinaryMeshIOJSBinding'],
      ['obj', 'itkOBJMeshIOJSBinding'],
      ['OBJ', 'itkOBJMeshIOJSBinding'],
      ['off', 'itkOFFMeshIOJSBinding'],
      ['OFF', 'itkOFFMeshIOJSBinding'],
      ['stl', 'itkSTLMeshIOJSBinding'],
      ['STL', 'itkSTLMeshIOJSBinding']
  ]);
  var extensionToIO$1 = extensionToIO;

  const MeshIOIndex = ['itkBYUMeshIOJSBinding', 'itkFreeSurferAsciiMeshIOJSBinding', 'itkFreeSurferBinaryMeshIOJSBinding', 'itkOBJMeshIOJSBinding', 'itkOFFMeshIOJSBinding', 'itkSTLMeshIOJSBinding', 'itkVTKPolyDataMeshIOJSBinding',];
  var MeshIOIndex$1 = MeshIOIndex;

  var axios$2 = {exports: {}};

  var bind$2 = function bind(fn, thisArg) {
    return function wrap() {
      var args = new Array(arguments.length);
      for (var i = 0; i < args.length; i++) {
        args[i] = arguments[i];
      }
      return fn.apply(thisArg, args);
    };
  };

  var bind$1 = bind$2;

  // utils is a library of generic helper functions non-specific to axios

  var toString = Object.prototype.toString;

  /**
   * Determine if a value is an Array
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is an Array, otherwise false
   */
  function isArray(val) {
    return toString.call(val) === '[object Array]';
  }

  /**
   * Determine if a value is undefined
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if the value is undefined, otherwise false
   */
  function isUndefined(val) {
    return typeof val === 'undefined';
  }

  /**
   * Determine if a value is a Buffer
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Buffer, otherwise false
   */
  function isBuffer(val) {
    return val !== null && !isUndefined(val) && val.constructor !== null && !isUndefined(val.constructor)
      && typeof val.constructor.isBuffer === 'function' && val.constructor.isBuffer(val);
  }

  /**
   * Determine if a value is an ArrayBuffer
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is an ArrayBuffer, otherwise false
   */
  function isArrayBuffer(val) {
    return toString.call(val) === '[object ArrayBuffer]';
  }

  /**
   * Determine if a value is a FormData
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is an FormData, otherwise false
   */
  function isFormData(val) {
    return (typeof FormData !== 'undefined') && (val instanceof FormData);
  }

  /**
   * Determine if a value is a view on an ArrayBuffer
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a view on an ArrayBuffer, otherwise false
   */
  function isArrayBufferView(val) {
    var result;
    if ((typeof ArrayBuffer !== 'undefined') && (ArrayBuffer.isView)) {
      result = ArrayBuffer.isView(val);
    } else {
      result = (val) && (val.buffer) && (val.buffer instanceof ArrayBuffer);
    }
    return result;
  }

  /**
   * Determine if a value is a String
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a String, otherwise false
   */
  function isString(val) {
    return typeof val === 'string';
  }

  /**
   * Determine if a value is a Number
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Number, otherwise false
   */
  function isNumber(val) {
    return typeof val === 'number';
  }

  /**
   * Determine if a value is an Object
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is an Object, otherwise false
   */
  function isObject(val) {
    return val !== null && typeof val === 'object';
  }

  /**
   * Determine if a value is a plain Object
   *
   * @param {Object} val The value to test
   * @return {boolean} True if value is a plain Object, otherwise false
   */
  function isPlainObject(val) {
    if (toString.call(val) !== '[object Object]') {
      return false;
    }

    var prototype = Object.getPrototypeOf(val);
    return prototype === null || prototype === Object.prototype;
  }

  /**
   * Determine if a value is a Date
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Date, otherwise false
   */
  function isDate(val) {
    return toString.call(val) === '[object Date]';
  }

  /**
   * Determine if a value is a File
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a File, otherwise false
   */
  function isFile(val) {
    return toString.call(val) === '[object File]';
  }

  /**
   * Determine if a value is a Blob
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Blob, otherwise false
   */
  function isBlob(val) {
    return toString.call(val) === '[object Blob]';
  }

  /**
   * Determine if a value is a Function
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Function, otherwise false
   */
  function isFunction(val) {
    return toString.call(val) === '[object Function]';
  }

  /**
   * Determine if a value is a Stream
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a Stream, otherwise false
   */
  function isStream(val) {
    return isObject(val) && isFunction(val.pipe);
  }

  /**
   * Determine if a value is a URLSearchParams object
   *
   * @param {Object} val The value to test
   * @returns {boolean} True if value is a URLSearchParams object, otherwise false
   */
  function isURLSearchParams(val) {
    return typeof URLSearchParams !== 'undefined' && val instanceof URLSearchParams;
  }

  /**
   * Trim excess whitespace off the beginning and end of a string
   *
   * @param {String} str The String to trim
   * @returns {String} The String freed of excess whitespace
   */
  function trim(str) {
    return str.trim ? str.trim() : str.replace(/^\s+|\s+$/g, '');
  }

  /**
   * Determine if we're running in a standard browser environment
   *
   * This allows axios to run in a web worker, and react-native.
   * Both environments support XMLHttpRequest, but not fully standard globals.
   *
   * web workers:
   *  typeof window -> undefined
   *  typeof document -> undefined
   *
   * react-native:
   *  navigator.product -> 'ReactNative'
   * nativescript
   *  navigator.product -> 'NativeScript' or 'NS'
   */
  function isStandardBrowserEnv() {
    if (typeof navigator !== 'undefined' && (navigator.product === 'ReactNative' ||
                                             navigator.product === 'NativeScript' ||
                                             navigator.product === 'NS')) {
      return false;
    }
    return (
      typeof window !== 'undefined' &&
      typeof document !== 'undefined'
    );
  }

  /**
   * Iterate over an Array or an Object invoking a function for each item.
   *
   * If `obj` is an Array callback will be called passing
   * the value, index, and complete array for each item.
   *
   * If 'obj' is an Object callback will be called passing
   * the value, key, and complete object for each property.
   *
   * @param {Object|Array} obj The object to iterate
   * @param {Function} fn The callback to invoke for each item
   */
  function forEach(obj, fn) {
    // Don't bother if no value provided
    if (obj === null || typeof obj === 'undefined') {
      return;
    }

    // Force an array if not already something iterable
    if (typeof obj !== 'object') {
      /*eslint no-param-reassign:0*/
      obj = [obj];
    }

    if (isArray(obj)) {
      // Iterate over array values
      for (var i = 0, l = obj.length; i < l; i++) {
        fn.call(null, obj[i], i, obj);
      }
    } else {
      // Iterate over object keys
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          fn.call(null, obj[key], key, obj);
        }
      }
    }
  }

  /**
   * Accepts varargs expecting each argument to be an object, then
   * immutably merges the properties of each object and returns result.
   *
   * When multiple objects contain the same key the later object in
   * the arguments list will take precedence.
   *
   * Example:
   *
   * ```js
   * var result = merge({foo: 123}, {foo: 456});
   * console.log(result.foo); // outputs 456
   * ```
   *
   * @param {Object} obj1 Object to merge
   * @returns {Object} Result of all merge properties
   */
  function merge(/* obj1, obj2, obj3, ... */) {
    var result = {};
    function assignValue(val, key) {
      if (isPlainObject(result[key]) && isPlainObject(val)) {
        result[key] = merge(result[key], val);
      } else if (isPlainObject(val)) {
        result[key] = merge({}, val);
      } else if (isArray(val)) {
        result[key] = val.slice();
      } else {
        result[key] = val;
      }
    }

    for (var i = 0, l = arguments.length; i < l; i++) {
      forEach(arguments[i], assignValue);
    }
    return result;
  }

  /**
   * Extends object a by mutably adding to it the properties of object b.
   *
   * @param {Object} a The object to be extended
   * @param {Object} b The object to copy properties from
   * @param {Object} thisArg The object to bind function to
   * @return {Object} The resulting value of object a
   */
  function extend(a, b, thisArg) {
    forEach(b, function assignValue(val, key) {
      if (thisArg && typeof val === 'function') {
        a[key] = bind$1(val, thisArg);
      } else {
        a[key] = val;
      }
    });
    return a;
  }

  /**
   * Remove byte order marker. This catches EF BB BF (the UTF-8 BOM)
   *
   * @param {string} content with BOM
   * @return {string} content value without BOM
   */
  function stripBOM(content) {
    if (content.charCodeAt(0) === 0xFEFF) {
      content = content.slice(1);
    }
    return content;
  }

  var utils$d = {
    isArray: isArray,
    isArrayBuffer: isArrayBuffer,
    isBuffer: isBuffer,
    isFormData: isFormData,
    isArrayBufferView: isArrayBufferView,
    isString: isString,
    isNumber: isNumber,
    isObject: isObject,
    isPlainObject: isPlainObject,
    isUndefined: isUndefined,
    isDate: isDate,
    isFile: isFile,
    isBlob: isBlob,
    isFunction: isFunction,
    isStream: isStream,
    isURLSearchParams: isURLSearchParams,
    isStandardBrowserEnv: isStandardBrowserEnv,
    forEach: forEach,
    merge: merge,
    extend: extend,
    trim: trim,
    stripBOM: stripBOM
  };

  var utils$c = utils$d;

  function encode(val) {
    return encodeURIComponent(val).
      replace(/%3A/gi, ':').
      replace(/%24/g, '$').
      replace(/%2C/gi, ',').
      replace(/%20/g, '+').
      replace(/%5B/gi, '[').
      replace(/%5D/gi, ']');
  }

  /**
   * Build a URL by appending params to the end
   *
   * @param {string} url The base of the url (e.g., http://www.google.com)
   * @param {object} [params] The params to be appended
   * @returns {string} The formatted url
   */
  var buildURL$2 = function buildURL(url, params, paramsSerializer) {
    /*eslint no-param-reassign:0*/
    if (!params) {
      return url;
    }

    var serializedParams;
    if (paramsSerializer) {
      serializedParams = paramsSerializer(params);
    } else if (utils$c.isURLSearchParams(params)) {
      serializedParams = params.toString();
    } else {
      var parts = [];

      utils$c.forEach(params, function serialize(val, key) {
        if (val === null || typeof val === 'undefined') {
          return;
        }

        if (utils$c.isArray(val)) {
          key = key + '[]';
        } else {
          val = [val];
        }

        utils$c.forEach(val, function parseValue(v) {
          if (utils$c.isDate(v)) {
            v = v.toISOString();
          } else if (utils$c.isObject(v)) {
            v = JSON.stringify(v);
          }
          parts.push(encode(key) + '=' + encode(v));
        });
      });

      serializedParams = parts.join('&');
    }

    if (serializedParams) {
      var hashmarkIndex = url.indexOf('#');
      if (hashmarkIndex !== -1) {
        url = url.slice(0, hashmarkIndex);
      }

      url += (url.indexOf('?') === -1 ? '?' : '&') + serializedParams;
    }

    return url;
  };

  var utils$b = utils$d;

  function InterceptorManager$1() {
    this.handlers = [];
  }

  /**
   * Add a new interceptor to the stack
   *
   * @param {Function} fulfilled The function to handle `then` for a `Promise`
   * @param {Function} rejected The function to handle `reject` for a `Promise`
   *
   * @return {Number} An ID used to remove interceptor later
   */
  InterceptorManager$1.prototype.use = function use(fulfilled, rejected, options) {
    this.handlers.push({
      fulfilled: fulfilled,
      rejected: rejected,
      synchronous: options ? options.synchronous : false,
      runWhen: options ? options.runWhen : null
    });
    return this.handlers.length - 1;
  };

  /**
   * Remove an interceptor from the stack
   *
   * @param {Number} id The ID that was returned by `use`
   */
  InterceptorManager$1.prototype.eject = function eject(id) {
    if (this.handlers[id]) {
      this.handlers[id] = null;
    }
  };

  /**
   * Iterate over all the registered interceptors
   *
   * This method is particularly useful for skipping over any
   * interceptors that may have become `null` calling `eject`.
   *
   * @param {Function} fn The function to call for each interceptor
   */
  InterceptorManager$1.prototype.forEach = function forEach(fn) {
    utils$b.forEach(this.handlers, function forEachHandler(h) {
      if (h !== null) {
        fn(h);
      }
    });
  };

  var InterceptorManager_1 = InterceptorManager$1;

  var utils$a = utils$d;

  var normalizeHeaderName$1 = function normalizeHeaderName(headers, normalizedName) {
    utils$a.forEach(headers, function processHeader(value, name) {
      if (name !== normalizedName && name.toUpperCase() === normalizedName.toUpperCase()) {
        headers[normalizedName] = value;
        delete headers[name];
      }
    });
  };

  /**
   * Update an Error with the specified config, error code, and response.
   *
   * @param {Error} error The error to update.
   * @param {Object} config The config.
   * @param {string} [code] The error code (for example, 'ECONNABORTED').
   * @param {Object} [request] The request.
   * @param {Object} [response] The response.
   * @returns {Error} The error.
   */
  var enhanceError$2 = function enhanceError(error, config, code, request, response) {
    error.config = config;
    if (code) {
      error.code = code;
    }

    error.request = request;
    error.response = response;
    error.isAxiosError = true;

    error.toJSON = function toJSON() {
      return {
        // Standard
        message: this.message,
        name: this.name,
        // Microsoft
        description: this.description,
        number: this.number,
        // Mozilla
        fileName: this.fileName,
        lineNumber: this.lineNumber,
        columnNumber: this.columnNumber,
        stack: this.stack,
        // Axios
        config: this.config,
        code: this.code,
        status: this.response && this.response.status ? this.response.status : null
      };
    };
    return error;
  };

  var enhanceError$1 = enhanceError$2;

  /**
   * Create an Error with the specified message, config, error code, request and response.
   *
   * @param {string} message The error message.
   * @param {Object} config The config.
   * @param {string} [code] The error code (for example, 'ECONNABORTED').
   * @param {Object} [request] The request.
   * @param {Object} [response] The response.
   * @returns {Error} The created error.
   */
  var createError$2 = function createError(message, config, code, request, response) {
    var error = new Error(message);
    return enhanceError$1(error, config, code, request, response);
  };

  var createError$1 = createError$2;

  /**
   * Resolve or reject a Promise based on response status.
   *
   * @param {Function} resolve A function that resolves the promise.
   * @param {Function} reject A function that rejects the promise.
   * @param {object} response The response.
   */
  var settle$1 = function settle(resolve, reject, response) {
    var validateStatus = response.config.validateStatus;
    if (!response.status || !validateStatus || validateStatus(response.status)) {
      resolve(response);
    } else {
      reject(createError$1(
        'Request failed with status code ' + response.status,
        response.config,
        null,
        response.request,
        response
      ));
    }
  };

  var utils$9 = utils$d;

  var cookies$1 = (
    utils$9.isStandardBrowserEnv() ?

    // Standard browser envs support document.cookie
      (function standardBrowserEnv() {
        return {
          write: function write(name, value, expires, path, domain, secure) {
            var cookie = [];
            cookie.push(name + '=' + encodeURIComponent(value));

            if (utils$9.isNumber(expires)) {
              cookie.push('expires=' + new Date(expires).toGMTString());
            }

            if (utils$9.isString(path)) {
              cookie.push('path=' + path);
            }

            if (utils$9.isString(domain)) {
              cookie.push('domain=' + domain);
            }

            if (secure === true) {
              cookie.push('secure');
            }

            document.cookie = cookie.join('; ');
          },

          read: function read(name) {
            var match = document.cookie.match(new RegExp('(^|;\\s*)(' + name + ')=([^;]*)'));
            return (match ? decodeURIComponent(match[3]) : null);
          },

          remove: function remove(name) {
            this.write(name, '', Date.now() - 86400000);
          }
        };
      })() :

    // Non standard browser env (web workers, react-native) lack needed support.
      (function nonStandardBrowserEnv() {
        return {
          write: function write() {},
          read: function read() { return null; },
          remove: function remove() {}
        };
      })()
  );

  /**
   * Determines whether the specified URL is absolute
   *
   * @param {string} url The URL to test
   * @returns {boolean} True if the specified URL is absolute, otherwise false
   */
  var isAbsoluteURL$1 = function isAbsoluteURL(url) {
    // A URL is considered absolute if it begins with "<scheme>://" or "//" (protocol-relative URL).
    // RFC 3986 defines scheme name as a sequence of characters beginning with a letter and followed
    // by any combination of letters, digits, plus, period, or hyphen.
    return /^([a-z][a-z\d\+\-\.]*:)?\/\//i.test(url);
  };

  /**
   * Creates a new URL by combining the specified URLs
   *
   * @param {string} baseURL The base URL
   * @param {string} relativeURL The relative URL
   * @returns {string} The combined URL
   */
  var combineURLs$1 = function combineURLs(baseURL, relativeURL) {
    return relativeURL
      ? baseURL.replace(/\/+$/, '') + '/' + relativeURL.replace(/^\/+/, '')
      : baseURL;
  };

  var isAbsoluteURL = isAbsoluteURL$1;
  var combineURLs = combineURLs$1;

  /**
   * Creates a new URL by combining the baseURL with the requestedURL,
   * only when the requestedURL is not already an absolute URL.
   * If the requestURL is absolute, this function returns the requestedURL untouched.
   *
   * @param {string} baseURL The base URL
   * @param {string} requestedURL Absolute or relative URL to combine
   * @returns {string} The combined full path
   */
  var buildFullPath$1 = function buildFullPath(baseURL, requestedURL) {
    if (baseURL && !isAbsoluteURL(requestedURL)) {
      return combineURLs(baseURL, requestedURL);
    }
    return requestedURL;
  };

  var utils$8 = utils$d;

  // Headers whose duplicates are ignored by node
  // c.f. https://nodejs.org/api/http.html#http_message_headers
  var ignoreDuplicateOf = [
    'age', 'authorization', 'content-length', 'content-type', 'etag',
    'expires', 'from', 'host', 'if-modified-since', 'if-unmodified-since',
    'last-modified', 'location', 'max-forwards', 'proxy-authorization',
    'referer', 'retry-after', 'user-agent'
  ];

  /**
   * Parse headers into an object
   *
   * ```
   * Date: Wed, 27 Aug 2014 08:58:49 GMT
   * Content-Type: application/json
   * Connection: keep-alive
   * Transfer-Encoding: chunked
   * ```
   *
   * @param {String} headers Headers needing to be parsed
   * @returns {Object} Headers parsed into an object
   */
  var parseHeaders$1 = function parseHeaders(headers) {
    var parsed = {};
    var key;
    var val;
    var i;

    if (!headers) { return parsed; }

    utils$8.forEach(headers.split('\n'), function parser(line) {
      i = line.indexOf(':');
      key = utils$8.trim(line.substr(0, i)).toLowerCase();
      val = utils$8.trim(line.substr(i + 1));

      if (key) {
        if (parsed[key] && ignoreDuplicateOf.indexOf(key) >= 0) {
          return;
        }
        if (key === 'set-cookie') {
          parsed[key] = (parsed[key] ? parsed[key] : []).concat([val]);
        } else {
          parsed[key] = parsed[key] ? parsed[key] + ', ' + val : val;
        }
      }
    });

    return parsed;
  };

  var utils$7 = utils$d;

  var isURLSameOrigin$1 = (
    utils$7.isStandardBrowserEnv() ?

    // Standard browser envs have full support of the APIs needed to test
    // whether the request URL is of the same origin as current location.
      (function standardBrowserEnv() {
        var msie = /(msie|trident)/i.test(navigator.userAgent);
        var urlParsingNode = document.createElement('a');
        var originURL;

        /**
      * Parse a URL to discover it's components
      *
      * @param {String} url The URL to be parsed
      * @returns {Object}
      */
        function resolveURL(url) {
          var href = url;

          if (msie) {
          // IE needs attribute set twice to normalize properties
            urlParsingNode.setAttribute('href', href);
            href = urlParsingNode.href;
          }

          urlParsingNode.setAttribute('href', href);

          // urlParsingNode provides the UrlUtils interface - http://url.spec.whatwg.org/#urlutils
          return {
            href: urlParsingNode.href,
            protocol: urlParsingNode.protocol ? urlParsingNode.protocol.replace(/:$/, '') : '',
            host: urlParsingNode.host,
            search: urlParsingNode.search ? urlParsingNode.search.replace(/^\?/, '') : '',
            hash: urlParsingNode.hash ? urlParsingNode.hash.replace(/^#/, '') : '',
            hostname: urlParsingNode.hostname,
            port: urlParsingNode.port,
            pathname: (urlParsingNode.pathname.charAt(0) === '/') ?
              urlParsingNode.pathname :
              '/' + urlParsingNode.pathname
          };
        }

        originURL = resolveURL(window.location.href);

        /**
      * Determine if a URL shares the same origin as the current location
      *
      * @param {String} requestURL The URL to test
      * @returns {boolean} True if URL shares the same origin, otherwise false
      */
        return function isURLSameOrigin(requestURL) {
          var parsed = (utils$7.isString(requestURL)) ? resolveURL(requestURL) : requestURL;
          return (parsed.protocol === originURL.protocol &&
              parsed.host === originURL.host);
        };
      })() :

    // Non standard browser envs (web workers, react-native) lack needed support.
      (function nonStandardBrowserEnv() {
        return function isURLSameOrigin() {
          return true;
        };
      })()
  );

  /**
   * A `Cancel` is an object that is thrown when an operation is canceled.
   *
   * @class
   * @param {string=} message The message.
   */
  function Cancel$3(message) {
    this.message = message;
  }

  Cancel$3.prototype.toString = function toString() {
    return 'Cancel' + (this.message ? ': ' + this.message : '');
  };

  Cancel$3.prototype.__CANCEL__ = true;

  var Cancel_1 = Cancel$3;

  var utils$6 = utils$d;
  var settle = settle$1;
  var cookies = cookies$1;
  var buildURL$1 = buildURL$2;
  var buildFullPath = buildFullPath$1;
  var parseHeaders = parseHeaders$1;
  var isURLSameOrigin = isURLSameOrigin$1;
  var createError = createError$2;
  var defaults$4 = defaults_1;
  var Cancel$2 = Cancel_1;

  var xhr = function xhrAdapter(config) {
    return new Promise(function dispatchXhrRequest(resolve, reject) {
      var requestData = config.data;
      var requestHeaders = config.headers;
      var responseType = config.responseType;
      var onCanceled;
      function done() {
        if (config.cancelToken) {
          config.cancelToken.unsubscribe(onCanceled);
        }

        if (config.signal) {
          config.signal.removeEventListener('abort', onCanceled);
        }
      }

      if (utils$6.isFormData(requestData)) {
        delete requestHeaders['Content-Type']; // Let the browser set it
      }

      var request = new XMLHttpRequest();

      // HTTP basic authentication
      if (config.auth) {
        var username = config.auth.username || '';
        var password = config.auth.password ? unescape(encodeURIComponent(config.auth.password)) : '';
        requestHeaders.Authorization = 'Basic ' + btoa(username + ':' + password);
      }

      var fullPath = buildFullPath(config.baseURL, config.url);
      request.open(config.method.toUpperCase(), buildURL$1(fullPath, config.params, config.paramsSerializer), true);

      // Set the request timeout in MS
      request.timeout = config.timeout;

      function onloadend() {
        if (!request) {
          return;
        }
        // Prepare the response
        var responseHeaders = 'getAllResponseHeaders' in request ? parseHeaders(request.getAllResponseHeaders()) : null;
        var responseData = !responseType || responseType === 'text' ||  responseType === 'json' ?
          request.responseText : request.response;
        var response = {
          data: responseData,
          status: request.status,
          statusText: request.statusText,
          headers: responseHeaders,
          config: config,
          request: request
        };

        settle(function _resolve(value) {
          resolve(value);
          done();
        }, function _reject(err) {
          reject(err);
          done();
        }, response);

        // Clean up request
        request = null;
      }

      if ('onloadend' in request) {
        // Use onloadend if available
        request.onloadend = onloadend;
      } else {
        // Listen for ready state to emulate onloadend
        request.onreadystatechange = function handleLoad() {
          if (!request || request.readyState !== 4) {
            return;
          }

          // The request errored out and we didn't get a response, this will be
          // handled by onerror instead
          // With one exception: request that using file: protocol, most browsers
          // will return status as 0 even though it's a successful request
          if (request.status === 0 && !(request.responseURL && request.responseURL.indexOf('file:') === 0)) {
            return;
          }
          // readystate handler is calling before onerror or ontimeout handlers,
          // so we should call onloadend on the next 'tick'
          setTimeout(onloadend);
        };
      }

      // Handle browser request cancellation (as opposed to a manual cancellation)
      request.onabort = function handleAbort() {
        if (!request) {
          return;
        }

        reject(createError('Request aborted', config, 'ECONNABORTED', request));

        // Clean up request
        request = null;
      };

      // Handle low level network errors
      request.onerror = function handleError() {
        // Real errors are hidden from us by the browser
        // onerror should only fire if it's a network error
        reject(createError('Network Error', config, null, request));

        // Clean up request
        request = null;
      };

      // Handle timeout
      request.ontimeout = function handleTimeout() {
        var timeoutErrorMessage = config.timeout ? 'timeout of ' + config.timeout + 'ms exceeded' : 'timeout exceeded';
        var transitional = config.transitional || defaults$4.transitional;
        if (config.timeoutErrorMessage) {
          timeoutErrorMessage = config.timeoutErrorMessage;
        }
        reject(createError(
          timeoutErrorMessage,
          config,
          transitional.clarifyTimeoutError ? 'ETIMEDOUT' : 'ECONNABORTED',
          request));

        // Clean up request
        request = null;
      };

      // Add xsrf header
      // This is only done if running in a standard browser environment.
      // Specifically not if we're in a web worker, or react-native.
      if (utils$6.isStandardBrowserEnv()) {
        // Add xsrf header
        var xsrfValue = (config.withCredentials || isURLSameOrigin(fullPath)) && config.xsrfCookieName ?
          cookies.read(config.xsrfCookieName) :
          undefined;

        if (xsrfValue) {
          requestHeaders[config.xsrfHeaderName] = xsrfValue;
        }
      }

      // Add headers to the request
      if ('setRequestHeader' in request) {
        utils$6.forEach(requestHeaders, function setRequestHeader(val, key) {
          if (typeof requestData === 'undefined' && key.toLowerCase() === 'content-type') {
            // Remove Content-Type if data is undefined
            delete requestHeaders[key];
          } else {
            // Otherwise add header to the request
            request.setRequestHeader(key, val);
          }
        });
      }

      // Add withCredentials to request if needed
      if (!utils$6.isUndefined(config.withCredentials)) {
        request.withCredentials = !!config.withCredentials;
      }

      // Add responseType to request if needed
      if (responseType && responseType !== 'json') {
        request.responseType = config.responseType;
      }

      // Handle progress if needed
      if (typeof config.onDownloadProgress === 'function') {
        request.addEventListener('progress', config.onDownloadProgress);
      }

      // Not all browsers support upload events
      if (typeof config.onUploadProgress === 'function' && request.upload) {
        request.upload.addEventListener('progress', config.onUploadProgress);
      }

      if (config.cancelToken || config.signal) {
        // Handle cancellation
        // eslint-disable-next-line func-names
        onCanceled = function(cancel) {
          if (!request) {
            return;
          }
          reject(!cancel || (cancel && cancel.type) ? new Cancel$2('canceled') : cancel);
          request.abort();
          request = null;
        };

        config.cancelToken && config.cancelToken.subscribe(onCanceled);
        if (config.signal) {
          config.signal.aborted ? onCanceled() : config.signal.addEventListener('abort', onCanceled);
        }
      }

      if (!requestData) {
        requestData = null;
      }

      // Send the request
      request.send(requestData);
    });
  };

  var utils$5 = utils$d;
  var normalizeHeaderName = normalizeHeaderName$1;
  var enhanceError = enhanceError$2;

  var DEFAULT_CONTENT_TYPE = {
    'Content-Type': 'application/x-www-form-urlencoded'
  };

  function setContentTypeIfUnset(headers, value) {
    if (!utils$5.isUndefined(headers) && utils$5.isUndefined(headers['Content-Type'])) {
      headers['Content-Type'] = value;
    }
  }

  function getDefaultAdapter() {
    var adapter;
    if (typeof XMLHttpRequest !== 'undefined') {
      // For browsers use XHR adapter
      adapter = xhr;
    } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
      // For node use HTTP adapter
      adapter = xhr;
    }
    return adapter;
  }

  function stringifySafely(rawValue, parser, encoder) {
    if (utils$5.isString(rawValue)) {
      try {
        (parser || JSON.parse)(rawValue);
        return utils$5.trim(rawValue);
      } catch (e) {
        if (e.name !== 'SyntaxError') {
          throw e;
        }
      }
    }

    return (encoder || JSON.stringify)(rawValue);
  }

  var defaults$3 = {

    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },

    adapter: getDefaultAdapter(),

    transformRequest: [function transformRequest(data, headers) {
      normalizeHeaderName(headers, 'Accept');
      normalizeHeaderName(headers, 'Content-Type');

      if (utils$5.isFormData(data) ||
        utils$5.isArrayBuffer(data) ||
        utils$5.isBuffer(data) ||
        utils$5.isStream(data) ||
        utils$5.isFile(data) ||
        utils$5.isBlob(data)
      ) {
        return data;
      }
      if (utils$5.isArrayBufferView(data)) {
        return data.buffer;
      }
      if (utils$5.isURLSearchParams(data)) {
        setContentTypeIfUnset(headers, 'application/x-www-form-urlencoded;charset=utf-8');
        return data.toString();
      }
      if (utils$5.isObject(data) || (headers && headers['Content-Type'] === 'application/json')) {
        setContentTypeIfUnset(headers, 'application/json');
        return stringifySafely(data);
      }
      return data;
    }],

    transformResponse: [function transformResponse(data) {
      var transitional = this.transitional || defaults$3.transitional;
      var silentJSONParsing = transitional && transitional.silentJSONParsing;
      var forcedJSONParsing = transitional && transitional.forcedJSONParsing;
      var strictJSONParsing = !silentJSONParsing && this.responseType === 'json';

      if (strictJSONParsing || (forcedJSONParsing && utils$5.isString(data) && data.length)) {
        try {
          return JSON.parse(data);
        } catch (e) {
          if (strictJSONParsing) {
            if (e.name === 'SyntaxError') {
              throw enhanceError(e, this, 'E_JSON_PARSE');
            }
            throw e;
          }
        }
      }

      return data;
    }],

    /**
     * A timeout in milliseconds to abort a request. If set to 0 (default) a
     * timeout is not created.
     */
    timeout: 0,

    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',

    maxContentLength: -1,
    maxBodyLength: -1,

    validateStatus: function validateStatus(status) {
      return status >= 200 && status < 300;
    },

    headers: {
      common: {
        'Accept': 'application/json, text/plain, */*'
      }
    }
  };

  utils$5.forEach(['delete', 'get', 'head'], function forEachMethodNoData(method) {
    defaults$3.headers[method] = {};
  });

  utils$5.forEach(['post', 'put', 'patch'], function forEachMethodWithData(method) {
    defaults$3.headers[method] = utils$5.merge(DEFAULT_CONTENT_TYPE);
  });

  var defaults_1 = defaults$3;

  var utils$4 = utils$d;
  var defaults$2 = defaults_1;

  /**
   * Transform the data for a request or a response
   *
   * @param {Object|String} data The data to be transformed
   * @param {Array} headers The headers for the request or response
   * @param {Array|Function} fns A single function or Array of functions
   * @returns {*} The resulting transformed data
   */
  var transformData$1 = function transformData(data, headers, fns) {
    var context = this || defaults$2;
    /*eslint no-param-reassign:0*/
    utils$4.forEach(fns, function transform(fn) {
      data = fn.call(context, data, headers);
    });

    return data;
  };

  var isCancel$1 = function isCancel(value) {
    return !!(value && value.__CANCEL__);
  };

  var utils$3 = utils$d;
  var transformData = transformData$1;
  var isCancel = isCancel$1;
  var defaults$1 = defaults_1;
  var Cancel$1 = Cancel_1;

  /**
   * Throws a `Cancel` if cancellation has been requested.
   */
  function throwIfCancellationRequested(config) {
    if (config.cancelToken) {
      config.cancelToken.throwIfRequested();
    }

    if (config.signal && config.signal.aborted) {
      throw new Cancel$1('canceled');
    }
  }

  /**
   * Dispatch a request to the server using the configured adapter.
   *
   * @param {object} config The config that is to be used for the request
   * @returns {Promise} The Promise to be fulfilled
   */
  var dispatchRequest$1 = function dispatchRequest(config) {
    throwIfCancellationRequested(config);

    // Ensure headers exist
    config.headers = config.headers || {};

    // Transform request data
    config.data = transformData.call(
      config,
      config.data,
      config.headers,
      config.transformRequest
    );

    // Flatten headers
    config.headers = utils$3.merge(
      config.headers.common || {},
      config.headers[config.method] || {},
      config.headers
    );

    utils$3.forEach(
      ['delete', 'get', 'head', 'post', 'put', 'patch', 'common'],
      function cleanHeaderConfig(method) {
        delete config.headers[method];
      }
    );

    var adapter = config.adapter || defaults$1.adapter;

    return adapter(config).then(function onAdapterResolution(response) {
      throwIfCancellationRequested(config);

      // Transform response data
      response.data = transformData.call(
        config,
        response.data,
        response.headers,
        config.transformResponse
      );

      return response;
    }, function onAdapterRejection(reason) {
      if (!isCancel(reason)) {
        throwIfCancellationRequested(config);

        // Transform response data
        if (reason && reason.response) {
          reason.response.data = transformData.call(
            config,
            reason.response.data,
            reason.response.headers,
            config.transformResponse
          );
        }
      }

      return Promise.reject(reason);
    });
  };

  var utils$2 = utils$d;

  /**
   * Config-specific merge-function which creates a new config-object
   * by merging two configuration objects together.
   *
   * @param {Object} config1
   * @param {Object} config2
   * @returns {Object} New object resulting from merging config2 to config1
   */
  var mergeConfig$2 = function mergeConfig(config1, config2) {
    // eslint-disable-next-line no-param-reassign
    config2 = config2 || {};
    var config = {};

    function getMergedValue(target, source) {
      if (utils$2.isPlainObject(target) && utils$2.isPlainObject(source)) {
        return utils$2.merge(target, source);
      } else if (utils$2.isPlainObject(source)) {
        return utils$2.merge({}, source);
      } else if (utils$2.isArray(source)) {
        return source.slice();
      }
      return source;
    }

    // eslint-disable-next-line consistent-return
    function mergeDeepProperties(prop) {
      if (!utils$2.isUndefined(config2[prop])) {
        return getMergedValue(config1[prop], config2[prop]);
      } else if (!utils$2.isUndefined(config1[prop])) {
        return getMergedValue(undefined, config1[prop]);
      }
    }

    // eslint-disable-next-line consistent-return
    function valueFromConfig2(prop) {
      if (!utils$2.isUndefined(config2[prop])) {
        return getMergedValue(undefined, config2[prop]);
      }
    }

    // eslint-disable-next-line consistent-return
    function defaultToConfig2(prop) {
      if (!utils$2.isUndefined(config2[prop])) {
        return getMergedValue(undefined, config2[prop]);
      } else if (!utils$2.isUndefined(config1[prop])) {
        return getMergedValue(undefined, config1[prop]);
      }
    }

    // eslint-disable-next-line consistent-return
    function mergeDirectKeys(prop) {
      if (prop in config2) {
        return getMergedValue(config1[prop], config2[prop]);
      } else if (prop in config1) {
        return getMergedValue(undefined, config1[prop]);
      }
    }

    var mergeMap = {
      'url': valueFromConfig2,
      'method': valueFromConfig2,
      'data': valueFromConfig2,
      'baseURL': defaultToConfig2,
      'transformRequest': defaultToConfig2,
      'transformResponse': defaultToConfig2,
      'paramsSerializer': defaultToConfig2,
      'timeout': defaultToConfig2,
      'timeoutMessage': defaultToConfig2,
      'withCredentials': defaultToConfig2,
      'adapter': defaultToConfig2,
      'responseType': defaultToConfig2,
      'xsrfCookieName': defaultToConfig2,
      'xsrfHeaderName': defaultToConfig2,
      'onUploadProgress': defaultToConfig2,
      'onDownloadProgress': defaultToConfig2,
      'decompress': defaultToConfig2,
      'maxContentLength': defaultToConfig2,
      'maxBodyLength': defaultToConfig2,
      'transport': defaultToConfig2,
      'httpAgent': defaultToConfig2,
      'httpsAgent': defaultToConfig2,
      'cancelToken': defaultToConfig2,
      'socketPath': defaultToConfig2,
      'responseEncoding': defaultToConfig2,
      'validateStatus': mergeDirectKeys
    };

    utils$2.forEach(Object.keys(config1).concat(Object.keys(config2)), function computeConfigValue(prop) {
      var merge = mergeMap[prop] || mergeDeepProperties;
      var configValue = merge(prop);
      (utils$2.isUndefined(configValue) && merge !== mergeDirectKeys) || (config[prop] = configValue);
    });

    return config;
  };

  var data = {
    "version": "0.23.0"
  };

  var VERSION = data.version;

  var validators$1 = {};

  // eslint-disable-next-line func-names
  ['object', 'boolean', 'number', 'function', 'string', 'symbol'].forEach(function(type, i) {
    validators$1[type] = function validator(thing) {
      return typeof thing === type || 'a' + (i < 1 ? 'n ' : ' ') + type;
    };
  });

  var deprecatedWarnings = {};

  /**
   * Transitional option validator
   * @param {function|boolean?} validator - set to false if the transitional option has been removed
   * @param {string?} version - deprecated version / removed since version
   * @param {string?} message - some message with additional info
   * @returns {function}
   */
  validators$1.transitional = function transitional(validator, version, message) {
    function formatMessage(opt, desc) {
      return '[Axios v' + VERSION + '] Transitional option \'' + opt + '\'' + desc + (message ? '. ' + message : '');
    }

    // eslint-disable-next-line func-names
    return function(value, opt, opts) {
      if (validator === false) {
        throw new Error(formatMessage(opt, ' has been removed' + (version ? ' in ' + version : '')));
      }

      if (version && !deprecatedWarnings[opt]) {
        deprecatedWarnings[opt] = true;
        // eslint-disable-next-line no-console
        console.warn(
          formatMessage(
            opt,
            ' has been deprecated since v' + version + ' and will be removed in the near future'
          )
        );
      }

      return validator ? validator(value, opt, opts) : true;
    };
  };

  /**
   * Assert object's properties type
   * @param {object} options
   * @param {object} schema
   * @param {boolean?} allowUnknown
   */

  function assertOptions(options, schema, allowUnknown) {
    if (typeof options !== 'object') {
      throw new TypeError('options must be an object');
    }
    var keys = Object.keys(options);
    var i = keys.length;
    while (i-- > 0) {
      var opt = keys[i];
      var validator = schema[opt];
      if (validator) {
        var value = options[opt];
        var result = value === undefined || validator(value, opt, options);
        if (result !== true) {
          throw new TypeError('option ' + opt + ' must be ' + result);
        }
        continue;
      }
      if (allowUnknown !== true) {
        throw Error('Unknown option ' + opt);
      }
    }
  }

  var validator$1 = {
    assertOptions: assertOptions,
    validators: validators$1
  };

  var utils$1 = utils$d;
  var buildURL = buildURL$2;
  var InterceptorManager = InterceptorManager_1;
  var dispatchRequest = dispatchRequest$1;
  var mergeConfig$1 = mergeConfig$2;
  var validator = validator$1;

  var validators = validator.validators;
  /**
   * Create a new instance of Axios
   *
   * @param {Object} instanceConfig The default config for the instance
   */
  function Axios$1(instanceConfig) {
    this.defaults = instanceConfig;
    this.interceptors = {
      request: new InterceptorManager(),
      response: new InterceptorManager()
    };
  }

  /**
   * Dispatch a request
   *
   * @param {Object} config The config specific for this request (merged with this.defaults)
   */
  Axios$1.prototype.request = function request(config) {
    /*eslint no-param-reassign:0*/
    // Allow for axios('example/url'[, config]) a la fetch API
    if (typeof config === 'string') {
      config = arguments[1] || {};
      config.url = arguments[0];
    } else {
      config = config || {};
    }

    config = mergeConfig$1(this.defaults, config);

    // Set config.method
    if (config.method) {
      config.method = config.method.toLowerCase();
    } else if (this.defaults.method) {
      config.method = this.defaults.method.toLowerCase();
    } else {
      config.method = 'get';
    }

    var transitional = config.transitional;

    if (transitional !== undefined) {
      validator.assertOptions(transitional, {
        silentJSONParsing: validators.transitional(validators.boolean),
        forcedJSONParsing: validators.transitional(validators.boolean),
        clarifyTimeoutError: validators.transitional(validators.boolean)
      }, false);
    }

    // filter out skipped interceptors
    var requestInterceptorChain = [];
    var synchronousRequestInterceptors = true;
    this.interceptors.request.forEach(function unshiftRequestInterceptors(interceptor) {
      if (typeof interceptor.runWhen === 'function' && interceptor.runWhen(config) === false) {
        return;
      }

      synchronousRequestInterceptors = synchronousRequestInterceptors && interceptor.synchronous;

      requestInterceptorChain.unshift(interceptor.fulfilled, interceptor.rejected);
    });

    var responseInterceptorChain = [];
    this.interceptors.response.forEach(function pushResponseInterceptors(interceptor) {
      responseInterceptorChain.push(interceptor.fulfilled, interceptor.rejected);
    });

    var promise;

    if (!synchronousRequestInterceptors) {
      var chain = [dispatchRequest, undefined];

      Array.prototype.unshift.apply(chain, requestInterceptorChain);
      chain = chain.concat(responseInterceptorChain);

      promise = Promise.resolve(config);
      while (chain.length) {
        promise = promise.then(chain.shift(), chain.shift());
      }

      return promise;
    }


    var newConfig = config;
    while (requestInterceptorChain.length) {
      var onFulfilled = requestInterceptorChain.shift();
      var onRejected = requestInterceptorChain.shift();
      try {
        newConfig = onFulfilled(newConfig);
      } catch (error) {
        onRejected(error);
        break;
      }
    }

    try {
      promise = dispatchRequest(newConfig);
    } catch (error) {
      return Promise.reject(error);
    }

    while (responseInterceptorChain.length) {
      promise = promise.then(responseInterceptorChain.shift(), responseInterceptorChain.shift());
    }

    return promise;
  };

  Axios$1.prototype.getUri = function getUri(config) {
    config = mergeConfig$1(this.defaults, config);
    return buildURL(config.url, config.params, config.paramsSerializer).replace(/^\?/, '');
  };

  // Provide aliases for supported request methods
  utils$1.forEach(['delete', 'get', 'head', 'options'], function forEachMethodNoData(method) {
    /*eslint func-names:0*/
    Axios$1.prototype[method] = function(url, config) {
      return this.request(mergeConfig$1(config || {}, {
        method: method,
        url: url,
        data: (config || {}).data
      }));
    };
  });

  utils$1.forEach(['post', 'put', 'patch'], function forEachMethodWithData(method) {
    /*eslint func-names:0*/
    Axios$1.prototype[method] = function(url, data, config) {
      return this.request(mergeConfig$1(config || {}, {
        method: method,
        url: url,
        data: data
      }));
    };
  });

  var Axios_1 = Axios$1;

  var Cancel = Cancel_1;

  /**
   * A `CancelToken` is an object that can be used to request cancellation of an operation.
   *
   * @class
   * @param {Function} executor The executor function.
   */
  function CancelToken(executor) {
    if (typeof executor !== 'function') {
      throw new TypeError('executor must be a function.');
    }

    var resolvePromise;

    this.promise = new Promise(function promiseExecutor(resolve) {
      resolvePromise = resolve;
    });

    var token = this;

    // eslint-disable-next-line func-names
    this.promise.then(function(cancel) {
      if (!token._listeners) return;

      var i;
      var l = token._listeners.length;

      for (i = 0; i < l; i++) {
        token._listeners[i](cancel);
      }
      token._listeners = null;
    });

    // eslint-disable-next-line func-names
    this.promise.then = function(onfulfilled) {
      var _resolve;
      // eslint-disable-next-line func-names
      var promise = new Promise(function(resolve) {
        token.subscribe(resolve);
        _resolve = resolve;
      }).then(onfulfilled);

      promise.cancel = function reject() {
        token.unsubscribe(_resolve);
      };

      return promise;
    };

    executor(function cancel(message) {
      if (token.reason) {
        // Cancellation has already been requested
        return;
      }

      token.reason = new Cancel(message);
      resolvePromise(token.reason);
    });
  }

  /**
   * Throws a `Cancel` if cancellation has been requested.
   */
  CancelToken.prototype.throwIfRequested = function throwIfRequested() {
    if (this.reason) {
      throw this.reason;
    }
  };

  /**
   * Subscribe to the cancel signal
   */

  CancelToken.prototype.subscribe = function subscribe(listener) {
    if (this.reason) {
      listener(this.reason);
      return;
    }

    if (this._listeners) {
      this._listeners.push(listener);
    } else {
      this._listeners = [listener];
    }
  };

  /**
   * Unsubscribe from the cancel signal
   */

  CancelToken.prototype.unsubscribe = function unsubscribe(listener) {
    if (!this._listeners) {
      return;
    }
    var index = this._listeners.indexOf(listener);
    if (index !== -1) {
      this._listeners.splice(index, 1);
    }
  };

  /**
   * Returns an object that contains a new `CancelToken` and a function that, when called,
   * cancels the `CancelToken`.
   */
  CancelToken.source = function source() {
    var cancel;
    var token = new CancelToken(function executor(c) {
      cancel = c;
    });
    return {
      token: token,
      cancel: cancel
    };
  };

  var CancelToken_1 = CancelToken;

  /**
   * Syntactic sugar for invoking a function and expanding an array for arguments.
   *
   * Common use case would be to use `Function.prototype.apply`.
   *
   *  ```js
   *  function f(x, y, z) {}
   *  var args = [1, 2, 3];
   *  f.apply(null, args);
   *  ```
   *
   * With `spread` this example can be re-written.
   *
   *  ```js
   *  spread(function(x, y, z) {})([1, 2, 3]);
   *  ```
   *
   * @param {Function} callback
   * @returns {Function}
   */
  var spread = function spread(callback) {
    return function wrap(arr) {
      return callback.apply(null, arr);
    };
  };

  /**
   * Determines whether the payload is an error thrown by Axios
   *
   * @param {*} payload The value to test
   * @returns {boolean} True if the payload is an error thrown by Axios, otherwise false
   */
  var isAxiosError = function isAxiosError(payload) {
    return (typeof payload === 'object') && (payload.isAxiosError === true);
  };

  var utils = utils$d;
  var bind = bind$2;
  var Axios = Axios_1;
  var mergeConfig = mergeConfig$2;
  var defaults = defaults_1;

  /**
   * Create an instance of Axios
   *
   * @param {Object} defaultConfig The default config for the instance
   * @return {Axios} A new instance of Axios
   */
  function createInstance(defaultConfig) {
    var context = new Axios(defaultConfig);
    var instance = bind(Axios.prototype.request, context);

    // Copy axios.prototype to instance
    utils.extend(instance, Axios.prototype, context);

    // Copy context to instance
    utils.extend(instance, context);

    // Factory for creating new instances
    instance.create = function create(instanceConfig) {
      return createInstance(mergeConfig(defaultConfig, instanceConfig));
    };

    return instance;
  }

  // Create the default instance to be exported
  var axios$1 = createInstance(defaults);

  // Expose Axios class to allow class inheritance
  axios$1.Axios = Axios;

  // Expose Cancel & CancelToken
  axios$1.Cancel = Cancel_1;
  axios$1.CancelToken = CancelToken_1;
  axios$1.isCancel = isCancel$1;
  axios$1.VERSION = data.version;

  // Expose all/spread
  axios$1.all = function all(promises) {
    return Promise.all(promises);
  };
  axios$1.spread = spread;

  // Expose isAxiosError
  axios$1.isAxiosError = isAxiosError;

  axios$2.exports = axios$1;

  // Allow use of default import syntax in TypeScript
  axios$2.exports.default = axios$1;

  var axios = axios$2.exports;

  // Load the Emscripten module in the browser in a WebWorker.
  //
  // baseUrl is usually taken from '../itkConfig.js', but a different value
  // could be passed.
  async function loadEmscriptenModuleWebWorker(moduleRelativePathOrURL, baseUrl) {
      let modulePrefix = null;
      if (typeof moduleRelativePathOrURL !== 'string') {
          modulePrefix = moduleRelativePathOrURL.href;
      }
      else if (moduleRelativePathOrURL.startsWith('http')) {
          modulePrefix = moduleRelativePathOrURL;
      }
      else {
          modulePrefix = `${baseUrl}/${moduleRelativePathOrURL}`;
      }
      if (modulePrefix.endsWith('.js')) {
          modulePrefix = modulePrefix.substring(0, modulePrefix.length - 3);
      }
      if (modulePrefix.endsWith('.wasm')) {
          modulePrefix = modulePrefix.substring(0, modulePrefix.length - 5);
      }
      // importScripts / UMD is required over dynamic ESM import until Firefox
      // adds worker dynamic import support:
      // https://bugzilla.mozilla.org/show_bug.cgi?id=1540913
      const wasmBinaryPath = `${modulePrefix}.umd.wasm`;
      const response = await axios.get(wasmBinaryPath, { responseType: 'arraybuffer' });
      const wasmBinary = response.data;
      const modulePath = `${modulePrefix}.umd.js`;
      importScripts(modulePath);
      const moduleBaseName = modulePrefix.replace(/.*\//, '');
      // @ts-ignore: error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'WorkerGlobalScope & typeof globalThis'.
      const wrapperModule = self[moduleBaseName];
      const emscriptenModule = wrapperModule({ wasmBinary });
      return emscriptenModule;
  }

  const IntTypes = {
      Int8: 'int8',
      UInt8: 'uint8',
      Int16: 'int16',
      UInt16: 'uint16',
      Int32: 'int32',
      UInt32: 'uint32',
      Int64: 'int64',
      UInt64: 'uint64',
      SizeValueType: 'uint64',
      IdentifierType: 'uint64',
      IndexValueType: 'int64',
      OffsetValueType: 'int64'
  };
  var IntTypes$1 = IntTypes;

  const FloatTypes = {
      Float32: 'float32',
      Float64: 'float64',
      SpacePrecisionType: 'float64'
  };
  var FloatTypes$1 = FloatTypes;

  const PixelTypes = {
      Unknown: 'Unknown',
      Scalar: 'Scalar',
      RGB: 'RGB',
      RGBA: 'RGBA',
      Offset: 'Offset',
      Vector: 'Vector',
      Point: 'Point',
      CovariantVector: 'CovariantVector',
      SymmetricSecondRankTensor: 'SymmetricSecondRankTensor',
      DiffusionTensor3D: 'DiffusionTensor3D',
      Complex: 'Complex',
      FixedArray: 'FixedArray',
      Array: 'Array',
      Matrix: 'Matrix',
      VariableLengthVector: 'VariableLengthVector',
      VariableSizeMatrix: 'VariableSizeMatrix'
  };
  var PixelTypes$1 = PixelTypes;

  class MeshType {
      constructor(dimension = 2, pointComponentType = FloatTypes$1.Float32, pointPixelComponentType = FloatTypes$1.Float32, pointPixelType = PixelTypes$1.Scalar, pointPixelComponents = 1, cellComponentType = IntTypes$1.Int32, cellPixelComponentType = FloatTypes$1.Float32, cellPixelType = PixelTypes$1.Scalar, cellPixelComponents = 1) {
          this.dimension = dimension;
          this.pointComponentType = pointComponentType;
          this.pointPixelComponentType = pointPixelComponentType;
          this.pointPixelType = pointPixelType;
          this.pointPixelComponents = pointPixelComponents;
          this.cellComponentType = cellComponentType;
          this.cellPixelComponentType = cellPixelComponentType;
          this.cellPixelType = cellPixelType;
          this.cellPixelComponents = cellPixelComponents;
      }
  }
  var MeshType$1 = MeshType;

  class Mesh {
      constructor(mt = new MeshType$1()) {
          this.mt = mt;
          this.name = 'Mesh';
          this.meshType = mt;
          this.name = 'Mesh';
          this.numberOfPoints = 0;
          this.points = null;
          this.numberOfPointPixels = 0;
          this.pointData = null;
          this.numberOfCells = 0;
          this.cellBufferSize = 0;
          this.cells = null;
          this.numberOfCellPixels = 0;
          this.cellData = null;
      }
  }
  var Mesh$1 = Mesh;

  function meshIOComponentToJSComponent(emscriptenModule, ioComponentType) {
      let componentType = null;
      switch (ioComponentType) {
          case emscriptenModule.IOComponentType.UNKNOWNCOMPONENTTYPE: {
              componentType = null;
              break;
          }
          case emscriptenModule.IOComponentType.UCHAR: {
              componentType = IntTypes$1.UInt8;
              break;
          }
          case emscriptenModule.IOComponentType.CHAR: {
              componentType = IntTypes$1.Int8;
              break;
          }
          case emscriptenModule.IOComponentType.USHORT: {
              componentType = IntTypes$1.UInt16;
              break;
          }
          case emscriptenModule.IOComponentType.SHORT: {
              componentType = IntTypes$1.Int16;
              break;
          }
          case emscriptenModule.IOComponentType.UINT: {
              componentType = IntTypes$1.UInt32;
              break;
          }
          case emscriptenModule.IOComponentType.INT: {
              componentType = IntTypes$1.Int32;
              break;
          }
          case emscriptenModule.IOComponentType.ULONG: {
              componentType = IntTypes$1.UInt64;
              break;
          }
          case emscriptenModule.IOComponentType.LONG: {
              componentType = IntTypes$1.Int64;
              break;
          }
          case emscriptenModule.IOComponentType.ULONGLONG: {
              componentType = IntTypes$1.UInt64;
              break;
          }
          case emscriptenModule.IOComponentType.LONGLONG: {
              componentType = IntTypes$1.Int64;
              break;
          }
          case emscriptenModule.IOComponentType.FLOAT: {
              componentType = FloatTypes$1.Float32;
              break;
          }
          case emscriptenModule.IOComponentType.DOUBLE: {
              componentType = FloatTypes$1.Float64;
              break;
          }
          default:
              throw new Error('Unknown IO component type');
      }
      return componentType;
  }

  function meshIOPixelTypeToJSPixelType(emscriptenModule, ioPixelType) {
      let pixelType = null;
      switch (ioPixelType) {
          case emscriptenModule.IOPixelType.UNKNOWNPIXELTYPE: {
              pixelType = PixelTypes$1.Unknown;
              break;
          }
          case emscriptenModule.IOPixelType.SCALAR: {
              pixelType = PixelTypes$1.Scalar;
              break;
          }
          case emscriptenModule.IOPixelType.RGB: {
              pixelType = PixelTypes$1.RGB;
              break;
          }
          case emscriptenModule.IOPixelType.RGBA: {
              pixelType = PixelTypes$1.RGBA;
              break;
          }
          case emscriptenModule.IOPixelType.OFFSET: {
              pixelType = PixelTypes$1.Offset;
              break;
          }
          case emscriptenModule.IOPixelType.VECTOR: {
              pixelType = PixelTypes$1.Vector;
              break;
          }
          case emscriptenModule.IOPixelType.POINT: {
              pixelType = PixelTypes$1.Point;
              break;
          }
          case emscriptenModule.IOPixelType.COVARIANTVECTOR: {
              pixelType = PixelTypes$1.CovariantVector;
              break;
          }
          case emscriptenModule.IOPixelType.SYMMETRICSECONDRANKTENSOR: {
              pixelType = PixelTypes$1.SymmetricSecondRankTensor;
              break;
          }
          case emscriptenModule.IOPixelType.DIFFUSIONTENSOR3D: {
              pixelType = PixelTypes$1.DiffusionTensor3D;
              break;
          }
          case emscriptenModule.IOPixelType.COMPLEX: {
              pixelType = PixelTypes$1.Complex;
              break;
          }
          case emscriptenModule.IOPixelType.FIXEDARRAY: {
              pixelType = PixelTypes$1.FixedArray;
              break;
          }
          case emscriptenModule.IOPixelType.ARRAY: {
              pixelType = PixelTypes$1.Array;
              break;
          }
          case emscriptenModule.IOPixelType.MATRIX: {
              pixelType = PixelTypes$1.Matrix;
              break;
          }
          case emscriptenModule.IOPixelType.VARIABLELENGTHVECTOR: {
              pixelType = PixelTypes$1.VariableLengthVector;
              break;
          }
          case emscriptenModule.IOPixelType.VARIABLESIZEMATRIX: {
              pixelType = PixelTypes$1.VariableSizeMatrix;
              break;
          }
          default:
              throw new Error('Unknown IO pixel type');
      }
      return pixelType;
  }

  function readMeshEmscriptenFSFile(emscriptenModule, filePath) {
      const meshIO = new emscriptenModule.ITKMeshIO();
      meshIO.SetFileName(filePath);
      if (!meshIO.CanReadFile(filePath)) {
          throw new Error('Could not read file: ' + filePath);
      }
      meshIO.ReadMeshInformation();
      const ioDimensions = meshIO.GetPointDimension();
      const ioPointComponentType = meshIO.GetPointComponentType();
      const pointComponentType = meshIOComponentToJSComponent(emscriptenModule, ioPointComponentType);
      if (pointComponentType === null) {
          throw Error('point component type cannot be unknown / null');
      }
      const ioCellComponentType = meshIO.GetCellComponentType();
      const cellComponentType = meshIOComponentToJSComponent(emscriptenModule, ioCellComponentType);
      if (cellComponentType === null) {
          throw Error('cell component type cannot be unknown / null');
      }
      const ioPointPixelComponentType = meshIO.GetPointPixelComponentType();
      let pointPixelComponentType = meshIOComponentToJSComponent(emscriptenModule, ioPointPixelComponentType);
      if (pointPixelComponentType === null) {
          // default
          pointPixelComponentType = FloatTypes$1.Float32;
      }
      const ioPointPixelType = meshIO.GetPointPixelType();
      let pointPixelType = meshIOPixelTypeToJSPixelType(emscriptenModule, ioPointPixelType);
      if (pointPixelType === null) {
          // default
          pointPixelType = PixelTypes$1.Scalar;
      }
      const pointPixelComponents = meshIO.GetNumberOfPointPixelComponents();
      const ioCellPixelComponentType = meshIO.GetCellPixelComponentType();
      let cellPixelComponentType = meshIOComponentToJSComponent(emscriptenModule, ioCellPixelComponentType);
      if (cellPixelComponentType === null) {
          // default
          cellPixelComponentType = FloatTypes$1.Float32;
      }
      const ioCellPixelType = meshIO.GetCellPixelType();
      let cellPixelType = meshIOPixelTypeToJSPixelType(emscriptenModule, ioCellPixelType);
      if (cellPixelType === null) {
          // default
          cellPixelType = PixelTypes$1.Scalar;
      }
      const cellPixelComponents = meshIO.GetNumberOfCellPixelComponents();
      const meshType = new MeshType$1(ioDimensions, pointComponentType, pointPixelComponentType, pointPixelType, pointPixelComponents, cellComponentType, cellPixelComponentType, cellPixelType, cellPixelComponents);
      const mesh = new Mesh$1(meshType);
      mesh.numberOfPoints = meshIO.GetNumberOfPoints();
      if (mesh.numberOfPoints > 0) {
          mesh.points = meshIO.ReadPoints();
      }
      mesh.numberOfCells = meshIO.GetNumberOfCells();
      if (mesh.numberOfCells > 0) {
          mesh.cellBufferSize = meshIO.GetCellBufferSize();
          mesh.cells = meshIO.ReadCells();
      }
      mesh.numberOfPointPixels = meshIO.GetNumberOfPointPixels();
      if (mesh.numberOfPointPixels > 0) {
          mesh.pointData = meshIO.ReadPointData();
      }
      mesh.numberOfCellPixels = meshIO.GetNumberOfCellPixels();
      if (mesh.numberOfCellPixels > 0) {
          mesh.cellData = meshIO.ReadCellData();
      }
      return mesh;
  }

  function meshJSComponentToIOComponent(emscriptenModule, componentType) {
      let ioComponentType = null;
      switch (componentType) {
          case null: {
              ioComponentType = emscriptenModule.IOComponentType.UNKNOWNCOMPONENTTYPE;
              break;
          }
          case IntTypes$1.UInt8: {
              ioComponentType = emscriptenModule.IOComponentType.UCHAR;
              break;
          }
          case IntTypes$1.Int8: {
              ioComponentType = emscriptenModule.IOComponentType.CHAR;
              break;
          }
          case IntTypes$1.UInt16: {
              ioComponentType = emscriptenModule.IOComponentType.USHORT;
              break;
          }
          case IntTypes$1.Int16: {
              ioComponentType = emscriptenModule.IOComponentType.SHORT;
              break;
          }
          case IntTypes$1.UInt32: {
              ioComponentType = emscriptenModule.IOComponentType.UINT;
              break;
          }
          case IntTypes$1.Int32: {
              ioComponentType = emscriptenModule.IOComponentType.INT;
              break;
          }
          case IntTypes$1.UInt64: {
              ioComponentType = emscriptenModule.IOComponentType.ULONGLONG;
              break;
          }
          case IntTypes$1.Int64: {
              ioComponentType = emscriptenModule.IOComponentType.LONGLONG;
              break;
          }
          case FloatTypes$1.Float32: {
              ioComponentType = emscriptenModule.IOComponentType.FLOAT;
              break;
          }
          case FloatTypes$1.Float64: {
              ioComponentType = emscriptenModule.IOComponentType.DOUBLE;
              break;
          }
          default:
              throw new Error('Unknown IO component type');
      }
      return ioComponentType;
  }

  function meshJSPixelTypeToIOPixelType(emscriptenModule, pixelType) {
      let ioPixelType = null;
      switch (pixelType) {
          case PixelTypes$1.Unknown: {
              ioPixelType = emscriptenModule.IOPixelType.UNKNOWNPIXELTYPE;
              break;
          }
          case PixelTypes$1.Scalar: {
              ioPixelType = emscriptenModule.IOPixelType.SCALAR;
              break;
          }
          case PixelTypes$1.RGB: {
              ioPixelType = emscriptenModule.IOPixelType.RGB;
              break;
          }
          case PixelTypes$1.RGBA: {
              ioPixelType = emscriptenModule.IOPixelType.RGBA;
              break;
          }
          case PixelTypes$1.Offset: {
              ioPixelType = emscriptenModule.IOPixelType.OFFSET;
              break;
          }
          case PixelTypes$1.Vector: {
              ioPixelType = emscriptenModule.IOPixelType.VECTOR;
              break;
          }
          case PixelTypes$1.Point: {
              ioPixelType = emscriptenModule.IOPixelType.POINT;
              break;
          }
          case PixelTypes$1.CovariantVector: {
              ioPixelType = emscriptenModule.IOPixelType.COVARIANTVECTOR;
              break;
          }
          case PixelTypes$1.SymmetricSecondRankTensor: {
              ioPixelType = emscriptenModule.IOPixelType.SYMMETRICSECONDRANKTENSOR;
              break;
          }
          case PixelTypes$1.DiffusionTensor3D: {
              ioPixelType = emscriptenModule.IOPixelType.DIFFUSIONTENSOR3D;
              break;
          }
          case PixelTypes$1.Complex: {
              ioPixelType = emscriptenModule.IOPixelType.COMPLEX;
              break;
          }
          case PixelTypes$1.FixedArray: {
              ioPixelType = emscriptenModule.IOPixelType.FIXEDARRAY;
              break;
          }
          case PixelTypes$1.Array: {
              ioPixelType = emscriptenModule.IOPixelType.ARRAY;
              break;
          }
          case PixelTypes$1.Matrix: {
              ioPixelType = emscriptenModule.IOPixelType.MATRIX;
              break;
          }
          case PixelTypes$1.VariableLengthVector: {
              ioPixelType = emscriptenModule.IOPixelType.VARIABLELENGTHVECTOR;
              break;
          }
          case PixelTypes$1.VariableSizeMatrix: {
              ioPixelType = emscriptenModule.IOPixelType.VARIABLESIZEMATRIX;
              break;
          }
          default:
              throw new Error('Unknown IO pixel type');
      }
      return ioPixelType;
  }

  function writeMeshEmscriptenFSFile(emscriptenModule, options, mesh, filePath) {
      const meshIO = new emscriptenModule.ITKMeshIO();
      meshIO.SetFileName(filePath);
      if (!meshIO.CanWriteFile(filePath)) {
          throw new Error('Could not write file: ' + filePath);
      }
      const dimension = mesh.meshType.dimension;
      meshIO.SetPointDimension(dimension);
      const pointIOComponentType = meshJSComponentToIOComponent(emscriptenModule, mesh.meshType.pointComponentType);
      if (pointIOComponentType === null) {
          throw Error('point io component type cannot be unknown / null');
      }
      meshIO.SetPointComponentType(pointIOComponentType);
      const cellIOComponentType = meshJSComponentToIOComponent(emscriptenModule, mesh.meshType.cellComponentType);
      if (cellIOComponentType === null) {
          throw Error('cell io component type cannot be unknown / null');
      }
      meshIO.SetCellComponentType(cellIOComponentType);
      const pointPixelIOComponentType = meshJSComponentToIOComponent(emscriptenModule, mesh.meshType.pointPixelComponentType);
      if (pointPixelIOComponentType === null) {
          throw Error('point pixel io component type cannot be unknown / null');
      }
      meshIO.SetPointPixelComponentType(pointPixelIOComponentType);
      const pointIOPixelType = meshJSPixelTypeToIOPixelType(emscriptenModule, mesh.meshType.pointPixelType);
      meshIO.SetPointPixelType(pointIOPixelType);
      meshIO.SetNumberOfPointPixelComponents(mesh.meshType.pointPixelComponents);
      const cellPixelIOComponentType = meshJSComponentToIOComponent(emscriptenModule, mesh.meshType.cellPixelComponentType);
      if (cellPixelIOComponentType === null) {
          throw Error('cell pixel io component type cannot be unknown / null');
      }
      meshIO.SetCellPixelComponentType(cellPixelIOComponentType);
      const cellIOPixelType = meshJSPixelTypeToIOPixelType(emscriptenModule, mesh.meshType.cellPixelType);
      meshIO.SetCellPixelType(cellIOPixelType);
      meshIO.SetNumberOfCellPixelComponents(mesh.meshType.cellPixelComponents);
      let useCompression = false;
      if (typeof options.useCompression !== 'undefined') {
          useCompression = options.useCompression;
      }
      let binaryFileType = false;
      if (typeof options.binaryFileType !== 'undefined') {
          binaryFileType = options.binaryFileType;
      }
      meshIO.SetUseCompression(useCompression);
      if (binaryFileType) {
          meshIO.SetFileType(emscriptenModule.FileType.BINARY);
      }
      else {
          meshIO.SetFileType(emscriptenModule.FileType.ASCII);
      }
      meshIO.SetByteOrder(emscriptenModule.ByteOrder.LittleEndian);
      meshIO.SetNumberOfPoints(mesh.numberOfPoints);
      if (mesh.numberOfPoints > 0) {
          meshIO.SetUpdatePoints(true);
      }
      meshIO.SetNumberOfPointPixels(mesh.numberOfPointPixels);
      if (mesh.numberOfPointPixels > 0) {
          meshIO.SetUpdatePointData(true);
      }
      meshIO.SetNumberOfCells(mesh.numberOfCells);
      if (mesh.numberOfCells > 0) {
          meshIO.SetUpdateCells(true);
      }
      meshIO.SetNumberOfCellPixels(mesh.numberOfCellPixels);
      meshIO.SetCellBufferSize(mesh.cellBufferSize);
      if (mesh.numberOfCellPixels > 0) {
          meshIO.SetUpdatePointData(true);
      }
      meshIO.WriteMeshInformation();
      if (mesh.numberOfPoints > 0) {
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          const numberOfBytes = mesh.points.length * mesh.points.BYTES_PER_ELEMENT;
          const dataPtr = emscriptenModule._malloc(numberOfBytes);
          const dataHeap = new Uint8Array(emscriptenModule.HEAPU8.buffer, dataPtr, numberOfBytes);
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          dataHeap.set(new Uint8Array(mesh.points.buffer));
          meshIO.WritePoints(dataHeap.byteOffset);
          emscriptenModule._free(dataHeap.byteOffset);
      }
      if (mesh.numberOfCells > 0) {
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          const numberOfBytes = mesh.cells.length * mesh.cells.BYTES_PER_ELEMENT;
          const dataPtr = emscriptenModule._malloc(numberOfBytes);
          const dataHeap = new Uint8Array(emscriptenModule.HEAPU8.buffer, dataPtr, numberOfBytes);
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          dataHeap.set(new Uint8Array(mesh.cells.buffer));
          meshIO.WriteCells(dataHeap.byteOffset);
          emscriptenModule._free(dataHeap.byteOffset);
      }
      if (mesh.numberOfPointPixels > 0) {
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          const numberOfBytes = mesh.pointData.length * mesh.pointData.BYTES_PER_ELEMENT;
          const dataPtr = emscriptenModule._malloc(numberOfBytes);
          const dataHeap = new Uint8Array(emscriptenModule.HEAPU8.buffer, dataPtr, numberOfBytes);
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          dataHeap.set(new Uint8Array(mesh.pointData.buffer));
          meshIO.WritePointData(dataHeap.byteOffset);
          emscriptenModule._free(dataHeap.byteOffset);
      }
      if (mesh.numberOfCellPixels > 0) {
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          const numberOfBytes = mesh.cellData.length * mesh.cellData.BYTES_PER_ELEMENT;
          const dataPtr = emscriptenModule._malloc(numberOfBytes);
          const dataHeap = new Uint8Array(emscriptenModule.HEAPU8.buffer, dataPtr, numberOfBytes);
          // @ts-expect-error: error TS2531: Object is possibly 'null'.
          dataHeap.set(new Uint8Array(mesh.cellData.buffer));
          meshIO.WriteCellData(dataHeap.byteOffset);
          emscriptenModule._free(dataHeap.byteOffset);
      }
      meshIO.Write();
  }

  // To cache loaded io modules
  const ioToModule = new Map();
  async function readMesh(input) {
      const extension = getFileExtension(input.name);
      const mountpoint = '/work';
      let io = null;
      if (mimeToIO$1.has(input.type)) {
          io = mimeToIO$1.get(input.type);
      }
      else if (extensionToIO$1.has(extension)) {
          io = extensionToIO$1.get(extension);
      }
      else {
          for (let idx = 0; idx < MeshIOIndex$1.length; ++idx) {
              let ioModule = null;
              const trialIO = MeshIOIndex$1[idx];
              if (ioToModule.has(trialIO)) {
                  ioModule = ioToModule.get(trialIO);
              }
              else {
                  ioToModule.set(trialIO, await loadEmscriptenModuleWebWorker(trialIO, input.config.meshIOUrl));
                  ioModule = ioToModule.get(trialIO);
              }
              const meshIO = new ioModule.ITKMeshIO();
              ioModule.fs_mkdirs(mountpoint);
              const filePath = `${mountpoint}/${input.name}`;
              ioModule.fs_writeFile(filePath, new Uint8Array(input.data));
              meshIO.SetFileName(filePath);
              if (meshIO.CanReadFile(filePath)) {
                  io = trialIO;
                  ioModule.fs_unlink(filePath);
                  break;
              }
              ioModule.fs_unlink(filePath);
          }
      }
      if (io === null) {
          ioToModule.clear();
          throw new Error('Could not find IO for: ' + input.name);
      }
      let ioModule = null;
      if (ioToModule.has(io)) {
          ioModule = ioToModule.get(io);
      }
      else {
          ioToModule.set(io, await loadEmscriptenModuleWebWorker(io, input.config.meshIOUrl));
          ioModule = ioToModule.get(io);
      }
      ioModule.fs_mkdirs(mountpoint);
      const filePath = `${mountpoint}/${input.name}`;
      ioModule.fs_writeFile(filePath, new Uint8Array(input.data));
      const mesh = readMeshEmscriptenFSFile(ioModule, filePath);
      ioModule.fs_unlink(filePath);
      const transferables = [];
      if (mesh.points) {
          transferables.push(mesh.points.buffer);
      }
      if (mesh.pointData) {
          transferables.push(mesh.pointData.buffer);
      }
      if (mesh.cells) {
          transferables.push(mesh.cells.buffer);
      }
      if (mesh.cellData) {
          transferables.push(mesh.cellData.buffer);
      }
      return new registerWebworker.TransferableResponse(mesh, transferables);
  }
  async function writeMesh(input) {
      const extension = getFileExtension(input.name);
      const mountpoint = '/work';
      let io = null;
      if (mimeToIO$1.has(input.type)) {
          io = mimeToIO$1.get(input.type);
      }
      else if (extensionToIO$1.has(extension)) {
          io = extensionToIO$1.get(extension);
      }
      else {
          for (let idx = 0; idx < MeshIOIndex$1.length; ++idx) {
              let ioModule = null;
              const trialIO = MeshIOIndex$1[idx];
              if (ioToModule.has(trialIO)) {
                  ioModule = ioToModule.get(trialIO);
              }
              else {
                  ioToModule.set(trialIO, await loadEmscriptenModuleWebWorker(trialIO, input.config.meshIOUrl));
                  ioModule = ioToModule.get(trialIO);
              }
              const meshIO = new ioModule.ITKMeshIO();
              const filePath = mountpoint + '/' + input.name;
              meshIO.SetFileName(filePath);
              if (meshIO.CanWriteFile(filePath)) {
                  io = trialIO;
                  break;
              }
          }
      }
      if (io === null) {
          ioToModule.clear();
          throw new Error('Could not find IO for: ' + input.name);
      }
      let ioModule = null;
      if (ioToModule.has(io)) {
          ioModule = ioToModule.get(io);
      }
      else {
          ioToModule.set(io, await loadEmscriptenModuleWebWorker(io, input.config.meshIOUrl));
          ioModule = ioToModule.get(io);
      }
      const filePath = `${mountpoint}/${input.name}`;
      ioModule.fs_mkdirs(mountpoint);
      writeMeshEmscriptenFSFile(ioModule, { useCompression: input.useCompression, binaryFileType: input.binaryFileType }, input.mesh, filePath);
      const writtenFile = ioModule.fs_readFile(filePath, { encoding: 'binary' });
      ioModule.fs_unlink(filePath);
      return new registerWebworker.TransferableResponse(writtenFile.buffer, [writtenFile.buffer]);
  }

  registerWebworker(async function (input) {
      if (input.operation === 'readMesh') {
          return readMesh(input);
      }
      else if (input.operation === 'writeMesh') {
          return writeMesh(input);
      }
      else {
          throw new Error('Unknown worker operation');
      }
  });

})();
